src/algebra/algebra/basic.lean

```diff
@@ -41,14 +41,41 @@ set_option extends_priority 200 /- control priority of
 `instance [algebra R A] : has_scalar R A` -/
 
 /--
-Given a commutative (semi)ring `R`, an `R`-algebra is a (possibly noncommutative)
-(semi)ring `A` endowed with a morphism of rings `R →+* A` which lands in the
-center of `A`.
-
-For convenience, this typeclass extends `has_scalar R A` where the scalar action must
-agree with left multiplication by the image of the structure morphism.
-
-Given an `algebra R A` instance, the structure morphism `R →+* A` is denoted `algebra_map R A`.
+Given a commutative (semi)ring `R`, there are two ways to define an `R`-algebra structure on a
+(possibly noncommutative) (semi)ring `A`:
+* By endowing `A` with a morphism of rings `R →+* A` denoted `algebra_map R A` which lands in the
+  center of `A`.
+* By requiring `A` be an `R`-module such that the action associates and commutes with multiplication
+  as `r • (a₁ * a₂) = (r • a₁) * a₂ = a₁ * (r • a₂)`.
+
+We define `algebra R A` as in a way that subsumes both definitions, by extending `has_scalar R A`
+and requiring that this scalar action `r • x` must agree with left multiplication by the image of
+the structure morphism `algebra_map R A r * x`.
+
+As a result, there are two ways to talk about an `R`-algebra `A` when `A` is a semiring:
+1. ```lean
+   variables [comm_semiring R] [semiring A]
+   variables [algebra R A]
+   ```
+2. ```lean
+   variables [comm_semiring R] [semiring A]
+   variables [module R A] [smul_comm_class R A A] [is_scalar_tower R A A]
+   ```
+The first approach implies the second via typeclass search; so any lemma stated with the second set
+of arguments will automatically apply to the first set. Typeclass search does not know that the
+second approach implies the first, but this can be shown with:
+```lean
+example {R A : Type*} [comm_semiring R] [semiring A]
+  [module R A] [smul_comm_class R A A] [is_scalar_tower R A A] : algebra R A :=
+algebra.of_module smul_mul_assoc mul_smul_comm
+```
+
+The advantage of the second approach is that `comm_semiring R`, `semiring A`, and `module R A` can
+all be relaxed independently; for instance, this allows us to:
+* Replace `semiring A` with `non_unital_non_assoc_semiring A` in order to describe Lie algebras
```

Comment:
```suggestion
* Replace `semiring A` with `non_unital_non_assoc_semiring A` in order to describe non-unital and/or non-associative algebras
```
I'm not sure I'd mention Lie algebras here since:

1. It might mislead somebody to believe this is actually how we define them
2. When I read it I think, "Wait, did I miss something? What about the other axioms Lie algebras satisfy?"

Reply:
I don't have strong feelings though; this is just a suggestion.

Reply:
Updated